src/cargo/core/workspace.rs

```diff
@@ -283,6 +283,10 @@ impl<'gctx> Workspace<'gctx> {
         ws.member_ids.insert(id);
         ws.default_members.push(ws.current_manifest.clone());
         ws.set_resolve_behavior();
+        // The find_root function is used here to traverse the directory tree and locate the root of the workspace.
```

Comment:
A bit worried about this change. The in-memory property isn't held for ephemeral workspaces after this patch. It is also a bit fragile because the workspace discovery logic is only guarded by [the other place](
https://github.com/rust-lang/cargo/blob/0a5cb8977a4393e9e814d9c049bb3e441c18f72b/src/cargo/core/workspace.rs#L1836-L1848), which is not immediately clear and hard to track.

I wonder if we really need this line here. `git_install_reads_workspace_manifest` seems to fail if we don't do `find_root`, but does cargo ever use `profile` or anything from there? Things we might want to figure out:

* What kind of workspace support `cargo install --git` current provide?
* Can we skip this and still find all members and inheritable fields?

Reply:
> but does cargo ever use `profile` or anything from there?

After this PR we can only check the manifest for errors. Before that, I belive that we could use any information defined by the workspace. For example lints etc.



> * What kind of workspace support `cargo install --git` current provide?

Before this PR, we just load the whole workspace and create a workspace based on the root manifest. 
After this PR, we create the workspace based on the specific package and we don't use the information from the root manifest. Only searched the root to see if there is an error in the manifest.




> * Can we skip this and still find all members and inheritable fields?

After this PR, we only checked the manifest, we don't respect workspace information anymore. 

For example:
If we try `cargo install --git https://github.com/hi-rustin/test-cargo-git-install bin-tool;`

Before this change:
```console
test-cargo-git-install on  master via 🦀 v1.77.2 
❯ cargo install --git https://github.com/hi-rustin/test-cargo-git-install bin-tool 
    Updating git repository `https://github.com/hi-rustin/test-cargo-git-install`
warning: virtual workspace defaulting to `resolver = "1"` despite one or more workspace members being on edition 2021 which implies `resolver = "2"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
note: for more details see https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions
  Installing bin-tool v0.1.0 (https://github.com/hi-rustin/test-cargo-git-install#d7a594e5)
   Compiling lib1 v0.1.0 (/Users/hi-rustin/.cargo/git/checkouts/test-cargo-git-install-fae8728302b9b2da/d7a594e/lib1)
   Compiling bin-tool v0.1.0 (/Users/hi-rustin/.cargo/git/checkouts/test-cargo-git-install-fae8728302b9b2da/d7a594e/bin-tool)
error: usage of an `unsafe` block
 --> bin-tool/src/main.rs:2:5
  |
2 | /     unsafe {
3 | |         println!("Hello, world!");
4 | |     }
  | |_____^
  |
  = note: requested on the command line with `-F unsafe-code`

warning: unnecessary `unsafe` block
 --> bin-tool/src/main.rs:2:5
  |
2 |     unsafe {
  |     ^^^^^^ unnecessary `unsafe` block
  |
  = note: `#[warn(unused_unsafe)]` on by default

warning: `bin-tool` (bin "bin-tool") generated 1 warning
error: could not compile `bin-tool` (bin "bin-tool") due to 1 previous error; 1 warning emitted
error: failed to compile `bin-tool v0.1.0 (https://github.com/hi-rustin/test-cargo-git-install#d7a594e5)`, intermediate artifacts can be found at `/var/folders/j1/7l6snwpx6svgqxh79bz_d27m0000gn/T/cargo-installbO7SHc`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

After this change:
```console
test-cargo-git-install on  master via 🦀 v1.77.2 
❯ ../cargo/target/debug/cargo install --git  https://github.com/hi-rustin/test-cargo-git-install bin-tool

    Updating git repository `https://github.com/hi-rustin/test-cargo-git-install`
  Installing bin-tool v0.1.0 (https://github.com/hi-rustin/test-cargo-git-install#d7a594e5)
   Compiling lib1 v0.1.0 (https://github.com/hi-rustin/test-cargo-git-install#d7a594e5)
   Compiling bin-tool v0.1.0 (https://github.com/hi-rustin/test-cargo-git-install#d7a594e5)
    Finished `release` profile [optimized] target(s) in 2.57s
  Installing /Users/hi-rustin/.cargo/bin/bin-tool
   Installed package `bin-tool v0.1.0 (https://github.com/hi-rustin/test-cargo-git-install#d7a594e5)` (executable `bin-tool`)
```

Reply:
From a workspace information perspective, I think this PR introduces some regression.

Reply:
> From a workspace information perspective, I think this PR introduces some regression.

workspace is a feature which should not be there as it complicates all optimizations? it is only there because CARGO_TARGET_DIR does not work very well, isn't it?

Reply:
workspace is already there so to move forward we need to first figure out the feature set `cargo install --git` offers. However, I agree with soloturn, we may need to step back a bit to re-evaludate the bug. 

Both workspace initialization and `CARGO_TARGET_DIR` are symptoms of the bug. From my previous investigation, the bug is in workspace, members are collected as from `PathSource`, hence `cargo install` recognizes the as path dependencies. In contrast, when using in `git` dependency, workspace members will be marked as from `GitSource`. I wonder if we should mark the entire workspace with the same source as the root one for `cargo install`. Thinking of nested packages <https://github.com/rust-lang/rfcs/pull/3452>, if we had that today, we might also need this, otherwise nested package will be marked as from local path and never get updated.

Reply:
> workspace is a feature which should not be there as it complicates all optimizations? it is only there because CARGO_TARGET_DIR does not work very well, isn't it?

I think it's hard to say the workspace feature complicates all optimizations. As my example in https://github.com/hi-rustin/test-cargo-git-install, respect workspace information will help us avoid inconsistency.

The problem is not caused by the CARGO_TARGET_DIR. If users set a CARGO_TARGET_DIR, then all compilation would happen in the CARGO_TARGET_DIR. 

The problem here is as @weihanglo said, we cannot rebuild the binary after the git repository is updated.

Reply:
> I wonder if we should mark the entire workspace with the same source as the root one for `cargo install`. Thinking of nested packages [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452), if we had that today, we might also need this, otherwise nested package will be marked as from local path and never get updated.

That sounds like a reasonable solution, I will take a look.

Reply:
> I wonder if we should mark the entire workspace with the same source as the root one for `cargo install`

I've looked at how we calculate the fingerprint of the path member. 

```rust
pub fn path_args(ws: &Workspace<'_>, unit: &Unit) -> (PathBuf, PathBuf) {
    let ws_root = ws.root();
    let src = match unit. target.src_path() {
        TargetSourcePath::Path(path) => path.to_path_buf(),
        TargetSourcePath::Metabuild => unit.pkg.manifest().metabuild_path(ws.target_dir()),
    };
    assert!(src.is_absolute());
    if unit.pkg.package_id().source_id().is_path() {
        if let Ok(path) = src.strip_prefix(ws_root) {
            return (path.to_path_buf(), ws_root.to_path_buf());
        }
    }
    (src, unit.pkg.root().to_path_buf())
}
```

It seems we always try to track the workspace as much as possible.

So I am not sure what you mean by `mark the entire workspace with the same source`. Can you explain it a little bit?

Reply:
When we read a dependency from a Git source, no matter it is a single package or workspace member, it always returns a `PackageId` with `SourceKind::Git`. Under the hood it is actually a `PathSource` assoicated with a Git `SourceId`. This helps Cargo track the actual source. 

https://github.com/rust-lang/cargo/blob/b9d913e532df8bb6e14a507376d9fbbcacdcc0b0/src/cargo/sources/git/source.rs#L360-L365

However, in `cargo install` we use workspace so we lost that  sticky fashion.